File: backend/src/Operum.API/Controllers/FieldsController.cs
using Microsoft.AspNetCore.Mvc;
using Operum.API.Controllers.Base;
using Operum.Model.DTOs.Fields.Requests;
using Operum.Service.Services.Fields;

namespace Operum.API.Controllers
{
    [ApiController]
    [Route("api/trackers/{trackerId}/[controller]")]
    public class FieldsController(IFieldsService fieldsService) : BaseController
    {
        [HttpGet]
        public async Task<IActionResult> GetFieldList([FromRoute] string trackerId)
        {
            return GetApiResponse(await fieldsService.GetFieldList(trackerId));
        }

        [HttpGet("{fieldId}")]
        public async Task<IActionResult> GetField([FromRoute] string trackerId, [FromRoute] string fieldId)
        {
            return GetApiResponse(await fieldsService.GetField(trackerId, fieldId));
        }

        [HttpPost]
        public async Task<IActionResult> CreateField([FromRoute] string trackerId, [FromBody] CreateFieldDto field)
        {
            return GetApiResponse(await fieldsService.CreateField(trackerId, field));
        }

        [HttpDelete("{fieldId}")]
        public async Task<IActionResult> DeleteField([FromRoute] string trackerId, [FromRoute] string fieldId)
        {
            return GetApiResponse(await fieldsService.DeleteField(trackerId, fieldId));
        }

        [HttpPut("reorder")]
        public async Task<IActionResult> ReorderFields([FromRoute] string trackerId, [FromBody] ReorderFieldsDto reorderFields)
        {
            return GetApiResponse(await fieldsService.ReorderFields(trackerId, reorderFields));
        }

        [HttpPut("{fieldId}")]
        public async Task<IActionResult> UpdateField([FromRoute] string trackerId, [FromRoute] string fieldId, [FromBody] UpdateFieldDto field)
        {
            return GetApiResponse(await fieldsService.UpdateField(trackerId, fieldId, field));
        }
    }
}

File: frontend/src/features/fields/api/fieldsController.ts
import api, { LONG_REQUEST_TIMEOUT_MS } from "../../../shared/api/api";
import { ApiResponse } from "../../../shared/types/ApiResponse";
import { CreateFieldDto } from "../types/CreateFieldDto";
import {
    ExtractFieldsDto,
    ExtractFieldsResultDto,
} from "../types/ExtractFieldsDto";
import { FieldDto } from "../types/FieldDto";
import { UpdateFieldDto } from "../types/UpdateFieldDto";

export const fieldsController = {
    getFields: async (trackerId: string): Promise<ApiResponse<FieldDto[]>> => {
        return await api.get(`/trackers/${trackerId}/fields`);
    },
    createField: async (
        trackerId: string,
        values: CreateFieldDto
    ): Promise<ApiResponse<FieldDto>> => {
        return await api.post(`/trackers/${trackerId}/fields`, values);
    },
    updateFieldOrder: async (
        trackerId: string,
        fieldIds: string[]
    ): Promise<ApiResponse> => {
        return await api.put(`/trackers/${trackerId}/fields/reorder`, {
            fieldIds,
        });
    },
    updateField: async (
        trackerId: string,
        fieldId: string,
        values: UpdateFieldDto
    ): Promise<ApiResponse> => {
        return await api.put(
            `/trackers/${trackerId}/fields/${fieldId}`,
            values
        );
    },
    deleteField: async (
        trackerId: string,
        fieldId: string
    ): Promise<ApiResponse> => {
        return await api.delete(`/trackers/${trackerId}/fields/${fieldId}`);
    },
    extractFields: async (
        trackerId: string,
        values: ExtractFieldsDto
    ): Promise<ApiResponse<ExtractFieldsResultDto>> => {
        return await api.post(
            `/trackers/${trackerId}/fields/extract`,
            values,
            { timeout: LONG_REQUEST_TIMEOUT_MS }
        );
    },
};
